presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java

```diff
@@ -759,6 +762,16 @@ public boolean isOrderByRedundant(OrderBy orderBy)
                         field -> ImmutableMap.of(toSchemaTableName(field.getOriginTable().get()), field.getOriginColumnName().get())));
     }
 
+    public boolean isRewritten()
```

Comment:
too generic names. you need to have it per query level
`registerQueryForViewOptimization(Query node) `

Use Deque to register, eg.
`    private final Deque<Query> queriesForView = new ArrayDeque<>();`